### ISSUE TITLE
GKE schedules pods like fluentd-gcp that prevent scale down

### ISSUE THREAD

**Issue by matti** [kubernetes/kubernetes#69696]
/kind bug

**What happened**:
Multiple nodepools:
default: 3x f1-micro
extra: autoscaling 0..9 f1-micro
worker: autoscaling 0..999 n1-standard-1 with taints to prevent GKE pods

GKE schedules system pods like fluentd-gcp so that "extra" never goes back to 0

**What you expected to happen**:
After all deployments were deleted the cluster should have scaled to:
default: 3x f1-micro
extra: 0
worker: 0

**How to reproduce it (as minimally and precisely as possible)**:
scale up the cluster and remove the resources

**Environment**:
- Kubernetes version (use `kubectl version`): v1.9.7-gke.6
- Cloud provider or hardware configuration: GKE by Google
- OS (e.g. from /etc/os-release): cos

**Comment by ProProgrammer** [comment]
Do we have any updates here?

**Comment by matti** [comment]
you can workaround this by setting a taint that prevents these pods from scheduling

**Comment by ogirardot** [comment]
@matti hi, can you elaborate, I think I have the same kind of issue...

**Comment by matti** [comment]
https://discuss.kubernetes.io/t/gke-scaled-down-nodes-wont-terminate/4699

**Comment by ariskay** [comment]
I do have the same issue, GKE creates three pods inside the new spawned node, all of them belongs to the kube-system namespace. The only way to have the cluster autoscale scale down is to drain the node. After a few minutes, the node pool scales to 0 nodes and the instance that was created also gets deleted.

I'm using the following to drain all the nodes in the pool so at least I can ensure they get deleted and don't stay for 

`for node in $(kubectl get nodes -l cloud.google.com/gke-nodepool=mypoolname -o=name); do
  kubectl drain "$node" --ignore-daemonsets;
done`

**Comment by ariskay** [comment]
> you can workaround this by setting a taint that prevents these pods from scheduling

Creating taints on the node pool and adding the toleration to the pods did the trick, there is a pod that doesn't get scheduled and now GKE is able to scale down the node pool.

**Comment by vitaly-karasik** [comment]
>Creating taints on the node pool and adding the toleration to the pods did the trick
@matti ,@ariskay, is this possible with a single node pool or I have to create a few node pools?

**Comment by matti** [comment]
works with single or multiple node pools. I don't understand your question.

**Comment by vitaly-karasik** [comment]
@matti  - i meant single [node] pool vs a few  pools.

**Comment by matti** [comment]
I understood that, but I don't understand the question still.

**Comment by vitaly-karasik** [comment]
I just saw in https://discuss.kubernetes.io/t/gke-scaled-down-nodes-wont-terminate/4699 that you're using a few pools as a workaround for a similar issue, so I'm curious if it's possible to use taint workaround with a single pool.

**Comment by matti** [comment]
Oh now I got it: nope, if you taint your single pool, the google managed pods can not run anywhere.

what you can do is to create a single preemptible f1-micro pool for google pods to be ashamed for their existence.

**Comment by vitaly-karasik** [comment]
Thank you, got it.
In fact, it seems that I fixed my cluster downscaling using  poddisruptionbudget for kube-system pods, based on https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-to-set-pdbs-to-enable-ca-to-move-kube-system-pods

**Comment by matti** [comment]
Can you share that config so that I understand that better?

**Comment by vitaly-karasik** [comment]
Sure:

kubectl create poddisruptionbudget kube-proxy-pdb --namespace=kube-system --selector=app=kube-proxy --max-unavailable 1
kubectl create poddisruptionbudget fluentd-gcp-pdb --namespace=kube-system --selector=app=fluentd-gcp --max-unavailable 1
kubectl create poddisruptionbudget prometheus-to-sd-pdb --namespace=kube-system --selector=app=prometheus-to-sd  --max-unavailable 1

**Comment by matti** [comment]
so by default maxunavailable is 0 and therefore the node pool can not scale to zero?

how many node pools do you have? a single pool can not scale to zero (?) -- just asking because you mentioned "single" nodepool earlier.

**Comment by vitaly-karasik** [comment]
There is one node pool for now, with min 2 and maximum 8 nodes.
 And it had 4 or 3 nodes (don't remember exactly), while I expected downscale to 2 nodes.

**Comment by matti** [comment]
Okay got it. Thanks for the PodDisruptionBudget tip!

**Comment by matti** [comment]
@vitaly-karasik aren't the labels:

```
component=kube-proxy
k8s-app=fluentd-gcp
k8s-app=prometheus-to-sd
```

**Comment by matti** [comment]
when I describe the poddisruptionbudget:
```
  Warning  CalculateExpectedPodCountFailed  31m (x2 over 32m)     controllermanager  Failed to calculate the number of expected pods: found no controllers for pod "kube-proxy-gke-hamina-pool-2-db15b2b8-nz3c"
  Warning  CalculateExpectedPodCountFailed  30m (x3 over 34m)     controllermanager  Failed to calculate the number of expected pods: found no controllers for pod "kube-proxy-gke-hamina-pool-2-d8bccb55-958z"
  Warning  NoControllers                    14m (x17 over 32m)    controllermanager  found no controllers for pod "kube-proxy-gke-hamina-pool-2-db15b2b8-nz3c"
```

**Comment by matti** [comment]
FYI: google seems to have changed fluentd etc pods so that they tolerate ALL taints, so workaround described earlier doesn't work.

**Comment by matti** [comment]
I'm 99.9% sure that PodDisruptionBudgets don't help at all - the only way currently is to delete the unwanted nodes.

**Comment by vitaly-karasik** [comment]
@matti, I have to agree with you - I was under the opinion that PodDisruptionBudgets helps, but I couldn't reproduce this behaviour.

**Comment by matti** [comment]
but - to me it looks like google might have fixed this: my cluster scaled down to match my workload. Not sure yet.

**Comment by bchazalet** [comment]
@matti I've applied your poddisruptionbudgets on my GKE cluster, and it does work for me too.

**Comment by matti** [comment]
@bchazalet I don't have the PDBs anymore and it still works for me.

**Comment by bchazalet** [comment]
@matti interesting. What version are running on?

**Comment by matti** [comment]
latest gke 12.7 or something like that.

**Comment by furlongce** [comment]
It's not working for me, and the 'tain/toleration' workaround doesn't seem to work either. Does anyone have any other suggestions?

**Comment by sebastiansirch** [comment]
We observed the same issue but it seems to be resolved in 1.14.7-gke.10. Nodes are not recreated after scaling the pool to 0.

**Comment by ninja-** [comment]
@sebastiansirch I updated to 1.14 from 1.13 and it didn't help :(

**Comment by ninja-** [comment]
kubectl create poddisruptionbudget kube-proxy-pdb --namespace=kube-system --selector=component=kube-proxy --max-unavailable 1
kubectl create poddisruptionbudget fluentd-gcp-pdb --namespace=kube-system --selector=k8s-app=fluentd-gcp --max-unavailable 1
kubectl create poddisruptionbudget prometheus-to-sd-pdb --namespace=kube-system --selector=k8s-app=prometheus-to-sd --max-unavailable 1

these PDBs don't help as well

**Comment by ninja-** [comment]
I double checked and looks like the 1.14 update does fix this (the update button only updated the master at first) :o even with no PDBs

**Comment by The-Fonz** [comment]
Having this same problem on 1.15.9. The taint workaround does work, the PodDisruptionBudget does not. This is really a pity as autoscaling is one of the best features of GKE, and this getting broken by *sytem* pods is horrible and costing us a lot of money + effort to manually drain pools.

Is there any other workaround or fix on the way?

**Comment by ninja-** [comment]
@The-Fonz idk, I had the same issues but now it works just fine out of the bx on 1.14-ish

**Comment by The-Fonz** [comment]
@ninja- Thanks for the reply. Did you re-create the cluster or just upgrade? I'm doing the upgrade, maybe that makes a difference?

**Comment by The-Fonz** [comment]
I deleted the fluentd and prometheus daemonsets from my k8s cluster and subsequently re-enabled stackdriver-kubernetes through the gcloud sdk. Additionally, I set the autoscaling profile to the [more aggressive "optimize-utilization"](https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-autoscaler#autoscaling_profiles). The former seems to have fixed the broken autoscaling for my default node pools (the pools with taints were already working correctly), the latter speeds up the autoscaling considerably.

Looks like this:

```console
user@pc:~$ kubectl delete daemonset -n kube-system --cascade fluentd-gcp-v3.1.1
user@pc:~$ kubectl delete daemonset -n kube-system --cascade prometheus-to-sd
user@pc:~$ gcloud beta container clusters update forecasting --enable-stackdriver-kubernetes
user@pc:~$ gcloud beta container clusters update forecasting --autoscaling-profile optimize-utilization
```

**Comment by danielyaa5** [comment]
These three pods are causing me not to be able to scale down 
```kubectl create poddisruptionbudget kube-proxy-pdb --namespace=kube-system --selector=app=kube-proxy --max-unavailable 1
kubectl create poddisruptionbudget fluentd-gcp-pdb --namespace=kube-system --selector=app=fluentd-gcp --max-unavailable 1
kubectl create poddisruptionbudget prometheus-to-sd-pdb --namespace=kube-system --selector=app=prometheus-to-sd --max-unavailable 1```, also when I add these pdbs I get `controllermanager  No matching pods found`

**Comment by hmeerlo** [comment]
Why o why doesn't Google just give us access to the autoscaler logs? This is a really big issue, wasting lots of money this way

**Comment by MarkPare** [comment]
Having same issue. I have a cluster used only for ci/cd that runs Jenkins with k8s plugin. When jobs are run, they typically require another node be spun up to run jenkins slave pod. This works fine, but the cluster never scales down when jobs finish. The node lingers with a few system pods running on it and I have to manually delete it.  I can't help but think this is low-priority for google because effectively they stand to get more money from me in status quo.

**Comment by benvdh** [comment]
I had a similar issue on one of our clusters with 2 node pools:
- default: 3 nodes (no autoscaling)
- etl: 0 - 5 nodes (autoscaling enabled)

Even though I have not been able to solve the issue completely yet (1 node always remains in the autoscaling pool, that were 2 nodes before I implemented pod disruption budgets), I have been able to resolve the issue with the NoControllers warning. Just posting it here so others might be able to experiment further with this.

Two things are important with regard to pod disruption budgets:
1. The example in the FAQ ([here](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-to-set-pdbs-to-enable-ca-to-move-kube-system-pods)) mentions the max-unavailable option. However, that option only applies to pods managed by one of the following 4 controllers ([source](https://kubernetes.io/docs/tasks/run-application/configure-pdb/#arbitrary-controllers-and-selectors)):
    - Deployment
    - ReplicationController
    - ReplicaSet
    - StatefulSet
2. In case you want to apply this to pods managed by some other type of controler (e.g. DaemonSets), you can use the min-available option([source](https://kubernetes.io/docs/tasks/run-application/configure-pdb/#arbitrary-controllers-and-selectors))

While exploring the pods that remained running, it turns out that both fluentd-gcp and prometheus-to-sd are managed by DaemonSets, hence you get the warning above, as a PDB with the max-unavailable option cannot work, as discussed above. Setting the option to min-available should at least get rid of the warning. My current cluster has the following pod-disruption-budgets set in the kube-system namespace and no longer produces the warning above, and keeps only one node running in the node pool:

![Screenshot from 2020-06-28 21-53-04](https://user-images.githubusercontent.com/4882376/85957069-781ccc00-b98a-11ea-9c61-afa5fd510679.png)

Only metrics-server and kube-proxy do not have a PDB yet. Metrics-server seems not to be recommended to be put under a pod disruption budget as per [here](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-to-set-pdbs-to-enable-ca-to-move-kube-system-pods), given the log loss one may suffer. Kube-proxy might be a next good candidate as discussed [here](https://github.com/kubernetes/autoscaler/issues/525).

I have no idea when I will have more time to test this, so feel free to have a go and post results here...

**Comment by benvdh** [comment]
I had some spare time left to do some more testing, and finally was able to resolve the issue with a combination of node taints, and pod disruption budgets. The node-taints already seem to prevent a lot of additional services from running on the nodes in my auto-scaling pool (e.g. metrics server, heapster etc.), and then only 3 services seem to remain:
- kube-proxy
- fluentd-gcp
- prometheus-to-sd

Here is the updated pdb configuration I used:

![Screenshot from 2020-06-30 13-44-15](https://user-images.githubusercontent.com/4882376/86122354-00f54e00-bad8-11ea-810c-da77cd51d27b.png)

**Comment by VengefulAncient** [comment]
We are also experiencing this issue as well with `1.16.13-gke.1`. Fortunately, simply switching the cluster's autoscaling profile to `optimize-utilization` seemed to have resolved this issue for now - we are using a cluster of 0-3 nodes for running Drone pipelines and it does scale back to 0 once the pipeline pods terminate.

**Comment by harsh4870** [comment]
@benvdh Can you please share the example **PDB** command you used for your configuration or YAML config ? 

Also how about the **metrics-server**? i am running it inside **kube-system** single replicas. i don't mind setting PDB for it also if it goes down for 1 min. 

i have 4 pools 3 are **non-scalable** and only **1** is scalable i was thinking to fix the **metrics-server** using **node selector** to non-scalable node it.

Heapster is already departed so don't mind if it goes down and schedules again on a new node.  

> I had some spare time left to do some more testing, and finally was able to resolve the issue with a combination of node taints, and pod disruption budgets. The node-taints already seem to prevent a lot of additional services from running on the nodes in my auto-scaling pool (e.g. metrics server, heapster etc.), and then only 3 services seem to remain:
> 
> * kube-proxy
> * fluentd-gcp
> * prometheus-to-sd
> 
> Here is the updated pdb configuration I used:

> 
> ![Screenshot from 2020-06-30 13-44-15](https://user-images.githubusercontent.com/4882376/86122354-00f54e00-bad8-11ea-810c-da77cd51d27b.png)

**Comment by benvdh** [comment]
@harsh4870 From the [Autoscaler FAQ](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-to-set-pdbs-to-enable-ca-to-move-kube-system-pods):
```bash
kubectl create poddisruptionbudget <pdb name> --namespace=kube-system --selector app=<app name> --max-unavailable 1`
```

So to implement the fluentd-gcp pod disruption above you would use:
```bash
kubectl create poddisruptionbudget fluentd-gcp-pod-disruption-budget --namespace=kube-system --selector k8s-app=fluentd-gcp --min-available 2
```
To discover the available labels you need for the selector parameter above, you can just do a `kubectl describe` on the running pods in the kube-system namespace.

With regard to metrics-server: I cannot remember that it was causing any issues in scaling down the node pool, at least I don't have any pdb's setup for them, which, as mentioned above, was also discouraged in the official docs.

**Comment by harsh4870** [comment]
Thanks for the quick response, got it. i got confused due to **Max unavailable NA**.

**Comment by fuine6** [comment]
I have the same problem in 1.20.9-gke.1001.

![image](https://user-images.githubusercontent.com/34824302/135217854-f2b47a90-d905-4e9e-a078-653fc0fbc5d7.png)

I tried two solutions.

Solution 1: **PDB**(Invalid at 1.20.9-gke.1001, valid at 1.17.17-gke.3700)

When I did not use **PDB**, **CA** complained about kube-dns unmovable.
```
{
   "noDecisionStatus":{
      "measureTime":"1632897393",
      "noScaleDown":{
         "nodes":[
            {
               "reason":{
                  "parameters":[
                     "kube-dns-b4f5c58c7-pwzg5"
                  ],
                  "messageId":"no.scale.down.node.pod.kube.system.unmovable"
               }
            }
         ]
      }
   }
}
```
Then I used this **PDB**.
```yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  labels:
    k8s-app: kube-dns
  name: kube-dns-pdb
  namespace: kube-system
spec:
  maxUnavailable: 1
  selector:
    matchLabels:
      k8s-app: kube-dns
```
**CA** complained about other things, but there is no more information, I don't know how to solve it.
```
{
   "jsonPayload":{
      "noDecisionStatus":{
         "noScaleDown":{
            "nodes":[
               {
                  "reason":{
                     "messageId":"no.scale.down.node.no.place.to.move.pods"
                  }
               }
            ]
         }
      }
   }
}
```

Solution 2: taint
I added taint to the app pool, then the **CA** scales correctly, but kube-dns began to complain that pods could not be scheduled.
![image](https://user-images.githubusercontent.com/34824302/135220635-b8be7dde-b1a3-41e3-9a93-4a54f1548490.png)
Is this safe to ignore?

Is there a better way?

**Comment by matti** [comment]
not safe to ignore - I have created a autoscaling pool without any taints just for these GKE pods. smallest machine type and preemptible.

all my pods have taint "googleStayAway: NO_EXECUTE" toleration

**Comment by fuine6** [comment]
> not safe to ignore - I have created a autoscaling pool without any taints just for these GKE pods. smallest machine type and preemptible.

Does this mean that the pool cannot be scale down to zero?

**Comment by matti** [comment]
that "for googles stupid pods" pool can not. gke wont function without those pods. all other pools can.

**Comment by giorgiomondauto** [comment]
Was it found a solution for this issue? Hard to believe that since 2018 this issue is still open and not yet solved

**Comment by jkh52** [comment]
This is a GKE node pool autoscaling question.

GKE-managed kube-system pods should generally not prevent cluster-autoscaler from scaling down. For example, I know that some system pods use `"cluster-autoscaler.kubernetes.io/safe-to-evict": "true"` per https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md. If you see specific system pods that appear to prevent, please file a bug with GKE including cluster version.

/cc @MaciekPytel @bskiba do you know of any GKE documentation with node pool configuration (taints, PDB, etc) that address this?

Note that with GKE Autopilot this issue is moot.

**Comment by matti** [comment]
https://github.com/kubernetes/kubernetes/issues/69696#issuecomment-929925497

^ I have been running GKE like this now for two years.

- have one pool with f1-micro / g1-small without any taints
- all other pools have a special taint that google's pods dont know about, so they can not park themselves on those pools.

--> gke can scale down to almost zero.

I have no idea why this wont get fixed, it is super annoying.

**Comment by DominikGM** [comment]
@jkh52 [Issue 181973707](https://issuetracker.google.com/issues/181973707) (at Google's issue-tracker) lists some system-pods which block down-scaling.

Note also [this comment from Nov, 2018](https://issuetracker.google.com/issues/119440195#comment3) (from another Google issue-tracker) that setting up PDBs should be part of customers where they can define the level of disruption they're ok with.

I've just come across this with `kube-dns`.

**Comment by matti** [comment]
makes absolutely no sense that GKE users need to create PDBs

during the last 3 years and 3 months there have been different pods that have prevented scaledown - I have noticed them and created new PDBs and removed old ones.

this is a ridicilous cat and mouse game.

**Comment by jprogramtech** [comment]
Do we have a solution for this issue? I've been working with my cluster for two years and have never had an issue with scale down now all of a sudden i'm getting scaling issues with 'Pod is blocking scale down because it’s a non-daemonset, non-mirrored, non-pdb-assigned kube-system pod' and "Pod is blocking scale down because it has local storage" taking a look at logs I see 'no.scale.down.node.pod.has.local.storage' 'metrics-server-v0.4.4-7d7cd5ddd5-cdz5r' and also for evict system.

**Comment by canove** [comment]
Same here: 
`Pod is blocking scale down because it’s a non-daemonset, non-mirrored, non-pdb-assigned kube-system pod`

```
reason: {
   messageId: "no.scale.down.node.pod.kube.system.unmovable"
   parameters: [
     0: "kube-dns-autoscaler-f4d55555-vgmdx"
]}
```

**Comment by FearlessHyena** [comment]
Autoscale being blocked by GKE's own pods is a bit absurd and hard to justify IMHO. Also it seems like the PDB workaround doesn't work anymore with the new GKE versions
There also doesn't seem to be any traction since this issue was created more than 3 years ago 
It seems it was assigned to @leilajal last year (apologise for the direct mention but would be really grateful if a contributor could please chime in 🙏 )

**Comment by imesh** [comment]
I think we could use the approach explained on the following FAQ page for moving `kube-dns` and `metrics-server` pods:
https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-to-set-pdbs-to-enable-ca-to-move-kube-system-pods

**Comment by FearlessHyena** [comment]
I've tried the PDB workaround and it doesn't work reliably
Also as per that documentation 
>Metrics Server is best left alone, as restarting it causes the loss of metrics for >1 minute, as well as metrics in dashboard from the last 15 minutes. Metrics Server downtime also means effective HPA downtime as it relies on metrics. Add PDB for it only if you're sure you don't mind.

So this workaround doesn't cover all cases (as many folks have commented earlier). The fact that these are system pods created automatically by GKE should also imply that they should be automatically managed by GKE and not cause a glaring problem with their own system

**Comment by matti** [comment]
No it doesnt.

**Comment by VengefulAncient** [comment]
It does for us. (GKE/1.21)

**Comment by red8888** [comment]
@VengefulAncient if its working for you can you confirm the following:

* Is your node pool also on 1.21? 
* Can you see if its setting "cluster-autoscaler.kubernetes.io/safe-to-evict" on those pods for you? 
* What scaling profile are you using? optimize-utilization ?

**Comment by VengefulAncient** [comment]
@red8888
1. Yes, it is.
2. No, I don't see that.
3. Yes, "Optimise utilisation".

**Comment by matti** [comment]
lool

**Comment by KimiJL** [comment]
/reopen /remove-lifecycle rotten

- Seeing this issue in Node Pool version 1.27.8
- Kube-dns and metrics-server preventing nodes from scaling down

In our use case, we run some heavy job pods on demand in these node pools and expect them to scale down after, but GKE managed kube-system pods are preventing CA scale down.

Reopening to see if there are new solutions / updates.

**Comment by ccope** [comment]
@KimiJL it doesn't look like the issue got reopened. I just got here via googling to resolve this problem. If you don't have the power to trigger actions, I can create a new issue and mention this one in it

**Comment by KimiJL** [comment]
@ccope that would be great! 👍

**Comment by agates4** [comment]
> @KimiJL it doesn't look like the issue got reopened. I just got here via googling to resolve this problem. If you don't have the power to trigger actions, I can create a new issue and mention this one in it

Hey @ccope do you mind creating that issue?

**Comment by agates4** [comment]
So actually, GKE's official recommendation is to create PDBs. Before creating PDBs, first understand whether a particular pod is actually preventing the scale down. 

GKE's autoscaling event logs **are** available for viewing.
For example, you can run this query yourself, and replace the relevant {VARIABLE} within the query:
https://console.cloud.google.com/logs
```
resource.type="k8s_cluster"
resource.labels.project_id="{MY_PROJECT}"
resource.labels.location="{MY_LOCATION}"
resource.labels.cluster_name="{MY_CLUSTER}"
logName="projects/{MY_PROJECT}/logs/container.googleapis.com%2Fcluster-autoscaler-visibility" severity>=DEFAULT
```

After viewing the logs, 
```json
{
insertId: "240d9a85-9149-4f78-80dd-18c901d1045a@a1"
jsonPayload: {
noDecisionStatus: {
measureTime: "1708913229"
noScaleDown: {
nodes: [
0: {
node: {
mig: {
name: "gke-n2-standard-32-ephemera-a2a84e84-grp"
nodepool: "n2-standard-32-ephemeral-node-pool-f478"
zone: "us-zz-b"
}
name: "gke-n2-standard-32-ephemera-a2a84e84-p8ls"
}
reason: {
messageId: "no.scale.down.node.pod.kube.system.unmovable"
parameters: [
0: "metrics-server-v0.5.2-545fc98df4-ngs5d"
]}}]
nodesTotalCount: 1
}}}
logName: "projects/my-project/logs/container.googleapis.com%2Fcluster-autoscaler-visibility"
receiveTimestamp: "2024-02-26T02:07:09.922923654Z"
resource: {
labels: {
cluster_name: "standard"
location: "us-zz"
project_id: "my-project"
}
type: "k8s_cluster"
}
timestamp: "2024-02-26T02:07:09.421874385Z"
}
```

I saw that the `metrics-server-v0.5.2-545fc98df4-ngs5d` pod was deployed to my ideally-0-scale node. 
I checked out the GKE docs to further inspect the `no.scale.down.node.pod.kube.system.unmovable` log:
https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-autoscaler-visibility
And I found that the recommended way to deal with these deployments which block scale down, is to create a PDB:
https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-to-set-pdbs-to-enable-ca-to-move-kube-system-pods

Before creating the PDB, I needed to find the right label with which to target the metrics pod.
`kubectl -n kube-system describe metrics-server-v0.5.2-545fc98df4-ngs5d`
Revealed that the correct labeler here is one of:
```
Labels:               k8s-app=metrics-server
                      pod-template-hash=545fc98df4
                      version=v0.5.2
```

And with this information, I created the right PDB:
`kubectl create poddisruptionbudget metrics-server-pdb --namespace=kube-system --selector=k8s-app=metrics-server --max-unavailable 1`

Inspecting the resultant PDB ensures that the label was chosen properly, and that the controller could find the pod:
`kubectl -n kube-system describe pdb metrics-server-pdb`
```
Status:
    Allowed disruptions:  1
    Current:              1
    Desired:              0
    Total:                1
Events:                   <none>
```

And a few minutes after creating this PDB, this node which had been running for a few days has finally automatically scaled down.
These results are from GKE cluster+node version: `1.26.12-gke.1111000`

This can be more succinctly documented, but I'm not sure this repository is the right place for it. In any case, if you're reading this, I hope this has been helpful and that this will save you time.

**Comment by FearlessHyena** [comment]
Thanks for the additional follow up @agates4 

The PDB workaround seems to be unreliable however and can break when Google makes any changes to their system pods between versions (the earlier posts will give a better overview on this)

Also this still feels like a workaround TBH so I think this issue needs to be reopened
These are all system pods that are automatically created by GKE so ideally this use case should also be automatically managed by GKE instead of requiring a workaround like this

**Comment by matti** [comment]
just create a new issue and see how google does not care

**Comment by FearlessHyena** [comment]
> just create a new issue and see how google does not care

yea I'm worried it'll just get ignored like this issue 😞

**Comment by matti** [comment]
@FearlessHyena strange that your name is Fearless, yet I see so much fear. Here you go #123989 my young padawan

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/kubernetes/issues/123989#issuecomment-2030492248

> This should be a ticket with cloud support, the Kubernetes project is not responsible for GKE and this issue tracker is not the right place for GKE support requests.

**Comment by BenTheElder** [comment]
If someone provides reference to a ticket I'll do my best to escalate internally in this case but please understand that this issue tracker is very busy and is NOT an appropriate place to reach out for product support and that is not the role of any of the contributors here including myself. The purpose of this issue tracker is to track issues with the open source project.

If you identify a bug in the open source project, please file an issue here. But if you have a problem with a vendor product, please contact the vendor directly.

Even those of us working on the open source project at a vendor generally are not employed to or in a position to manage support cases directly versus being engaged by support teams to handle product bugs for the components we own internally. Filing a bug here does not get through to the right people and we cannot file support cases for you. Everyone working here wants you to have a good experience using Kubernetes but we are not in a position to manage vendor specific issues.
